### ISSUE TITLE
Enable easier upgrades across multiple versions of Sourcegraph

### ISSUE THREAD

**Issue by beyang** [sourcegraph/sourcegraph#10144]
Problem: Many Sourcegraph customers are several versions behind or more. It is currently painful to upgrade Sourcegraph across multiple versions, because:
* Upgrades must proceed one minor version at a time, because we don't guarantee forward compatibility if versions are skipped
* For Kubernetes deployments, each single version upgrade is painful, because it frequently requires resolving merge conflicts.

https://github.com/sourcegraph/sourcegraph/issues/10100 will help address the second point. The first remains an issue.

Strawman proposal: Upgrades require migrating any data with persistent storage. This currently includes:
* Postgres
* Redis (redis-cache and redis-store)
* Zoekt index
* Gitserver filesystem layout

In theory, could we do the following?
* Record the storage version for each of these services (i.e., the last Sourcegraph version to update the storage format)
* Each of these services would run a `migrate` package on startup that reads the last storage version and runs the appropriate migrations, similar in spirit to https://github.com/golang-migrate/migrate, which we already use for Postgres. The `migrate` package should be as simple as possible. E.g., if redis-store is too many versions behind, just delete everything (clearing all user sessions).

This would, in theory, make it possible to upgrade from any previous version of Sourcegraph to the latest version.

**Comment by tsenart** [comment]
> Record the storage version for each of these services (i.e., the last Sourcegraph version to update the storage format)

We already have code in place that would allow for this easily, which was introduced in https://github.com/sourcegraph/sourcegraph/pull/8157.

However, I don't immediately see how we can deprecate code gradually if we allow upgrades from any version. cc @keegancsmith, please comment.

**Comment by keegancsmith** [comment]
yeah we can't allow it from any version. For example we have code based migrations that happen. An example of a complicated one was taking repos which were disabled and updating the external service configuration to exclude them. This is not possible to represent. We used to have a "2 version" upgrade policy. I think an upgrade policy that is more specific may make sense:

1. Deprecated APIs can be removed in the next release
2. Data compatibility for 3 versions. (or just 2 for simplicity of automatically testing).

If a customer is on a version that is N versions old, then they would only need to do N / 3 upgrades.

Kubernetes being a pain to upgrade is orthogonal to this and should be addressed even if we could skip versions. To me that is the higher priority issue here.

**Comment by slimsag** [comment]
Related (but not a complete solution for this): https://github.com/sourcegraph/sourcegraph/issues/9056

**Comment by pecigonzalo** [comment]
@efritz I believe this is in the realm of the work you were doing for out-of-band migrations. I think this is a bit of a problem across all teams that needs to implemented for the different services we have.
